### DISCUSSION TITLE
INA219 increasing Amp resolution

### DISCUSSION THREAD

**Discussion by docaoc5716** [arendst/Tasmota#16830] [General]
Hi, I am new to this discussion group and I have a question. 
I am using a INA219 to read a 4-20 mA signal coming from a level sensor. 
The level sensor finds the amount of fuel in a 1500 liter home heating fuel tank. 
I would like to have more resolution in the tasmota current reading, presently it displays an amp reading of 3 decimal places. 
What can I do to display 0.0000 or 0.00000 amps? 
AmpRes is set to 3. 
sensor13 is set to 1. 
tasmota V 11.0.0
Using NodeMCU ESP8266

Thanks for the amazing Tasmota software

**Comment by barbudor** [comment]
What is your actual shunt resistor value ? Is that the standard 0.1 ohm (R100) that is provided o most chinese INA219 modules ?

If so you can fool the driver to let it belive you have a 100 ohm resistor (x1000 larger than the one you actually have) using `Sensor13 15`  and instead of displaying `0.004A` to ` 0.020 A`  it will display `4.000 A` to `20.000 A`

**Reply by docaoc5716** [reply]
Hi it is the standard resistor that comes with the Chinese module. I will try your solution and let you know how I get on soon
thanks for replying so quickly.

**Reply by docaoc5716** [reply]
Hi barbudor
Tried setting sensor 13 to 15 and the current reading is 0.000 amps. I changed it back to Sensor13 1 and is reads the original 0.017 amps. So it seems like it didn’t work.

**Reply by barbudor** [reply]
Need to check, should have worked

**Reply by barbudor** [reply]
Sorry, my math were rusty this afternoon
You need to fool the driver with a smaller shunt resistor, not a larger one.
But the smallest you can do is 0.001 ohm, so 100 times smaller than the original => `Sensor13 10` will give a *100 to the values so will be 0.400A to 2.000 A
That not as nice as what I expected but still give you 2 more digits

**Reply by docaoc5716** [reply]
No could not get 15 to work, changed sensor13 to 11 now the amp reading is 0.171 amps so  I did get an extra digit. the amp reading fluctuates a little bit 0.005 but I can average the readings in node red, the level sensor is a $400 flowline part it read steady when it was connected to a PLC and when measured with a meter so I presume the $4 INA219 may be causing the fluctuations. At least now I can get  a 9.3 liter step in the fuel level as opposed to 93 liter step (1500L/16mA).

**Reply by docaoc5716** [reply]
Yes sensor13 10 works sorry i was testing and typing my results and didnt notice your last message. I am happy with the reading now thanks for all your help barbudor.

**Reply by barbudor** [reply]
I made a change to the INA219 driver which is now in the development branch v12.1.1.6 (https://github.com/arendst/Tasmota/pull/16837)
If you use `Sensor13 0, 0.0001` you will get 4.000A to 20.000A which may be easier to read
Note that this setting won't be saved in flash so you need to have a rule to set it back at every reboot:
```
Rule1 on system#boot do Sensor13 0, 0.00001 endon
Rule1 1
```

**Reply by docaoc5716** [reply]
Hi
barbudor, I used the rule and then typed Rule1 1, but now I am getting an amp reading of 0.017,
the same as before I posed the question yesterday.
this is what I get when I do a rule1 command.

  20:52:21.377 MQT: stat/Tank_Level/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":43,"Free":468,"Rules":"on system#boot do Sensor13 0, 0.00001 endon"}}

This is the result of a sensor13 command.

21:04:49.219 CMD: sensor13
21:04:49.225 MQT: stat/Tank_Level/RESULT = {"Sensor13":0}

Do I need to upgrade the Tasmota version today?

**Reply by barbudor** [reply]
My change was added to Tasmota this morning
So yes you need to upgrade to the latest tasmota development version from today

**Comment by docaoc5716** [comment]
Upgraded the firmwear and I am getting a current value that is factor 10^5 greater than before, which is wonderful. 
"Current":166.000,. 
Just curious, why do you think everything after the decimal point is a zero, is it a limitation of the INA219 or the  NodeMCU ESP8266. According to The Level probe datasheet https://www.flowline.com/_data_sheet_and_manuals/current/EchoPod_DL14_Data_Sheet.pdf (Flowline Echopod DL14) has a resolution of 0.5mm over a span of 1.2 M.
(1200/0.5 = 2400 steps) 
(16mA/2400 =.00667mA per step)

I know I am getting a bit off topic with the level probe since this is a Tasmota discussion forum and not a Flowline user group. So barbudor thank you very much for all your help I really appreciate it. 

![tank level](https://user-images.githubusercontent.com/115884373/196066340-f39188ba-d096-48f3-84a2-dae0399b16d0.png)

**Reply by barbudor** [reply]
The smallest step on INA219 is given by the LSB of the shunt ADC which is 10uV
As per ohm law, this means that through a resistor of 0.1 ohm, the smallest step is 0.1mA
So you should have 1 valid digit after the dot.
Some averaging is enabled by default but I don't know the impact.

With the default 0.1 ohm, the current range of the ina219 is 0 to 3.2 A so much larger than your area of interrest
If you can change the resistor for a 1 ohm or even 10 ohm you could drastically increase the number of digits. But those digits would be meaning full only if you have a high accuracy resistor like 0.1%

**Reply by barbudor** [reply]
I don't understand why you get 166A
You are supposed to have a 4-20mA sensor which with the resistor tweak should be represented by Tasmota as 4-20A

**Reply by docaoc5716** [reply]
Would I get a reading of 16.6 amps if in rule1 Sensor13 0 was followed by 0.001 rather than 0.0001  . I’ll will source a few low tolerance 1 and 10 ohm shunt resistors and experiment with another INA219 and a 4-20 mA current source. I should mention that the fuel tank being monitored is located about 4800 Km from me in another country.

**Reply by barbudor** [reply]
There's 2 values to consider:
- the real resistor to get the best from the ina219
- the resistor value you'll give to Tasmota to have as much digits as possible


The Ina measure the current by measuring the voltage across the shunt resistor.
The max voltage across the shunt resistor is 320mV
So you need to be as close as possible to 320mV when  20mA goes through the resistor.
As per ohm law, V = R*I so with 10 ohm * 20mA you will get 200mV.
So you should get a 10ohm with the best accuracy as possible or get a high accuracy ohmeter to know the exact resistor value.

Now if you tell tasmota that you have such a 10 ohm resistor, it will display 0.004 to 0.020 despite getting much better value from the Ina.

So that the tweak to tell tasmota that you have a resistor that is 1000 smaller than your real resistor to have values shifted from mA to A.
For a real resistor of 10 ohm use 'Sensor13 0, 0.01'

I hope this is clear

**Reply by docaoc5716** [reply]
Are you saying that the best resolution is 0.1mA with a 0.1 ohm resistor and with a 10 ohm resistor a resolution of .001mA is possible?

**Reply by barbudor** [reply]
Officially thé résolution of the shunt ADC is 10uV
So yes, 0.1mA over 0.1 ohm and 1uA over 10 ohm
But that's theorical and also depends on the accuracy of the resistor.
But still you'll better numbers with 10 ohms even if you cant trust the last digits

**Reply by docaoc5716** [reply]
I appreciate all your help barbudor. Thanks. Is there a protocol to end the discussion and indicate a satisfactory solution to my original question?

**Reply by barbudor** [reply]
You're welcome 

There is no real closure on a discussion but you still select the most appropriate message and tag is as the answer

**Reply by diverjoe** [reply]
This has been a wonderfully informative discussion (says the fly on the wall) for me as well.  Thank  you barbudor!
